### ISSUE TITLE
Type checker fails distinguish different types that have the same name

### ISSUE THREAD

**Issue by peterwvj** [overturetool/overture#488]
Type checker fails distinguish different types that have the same name. To demonstrate this, consider the spec below (the errors reported by the type checker have been added as comments to the spec):

~~~
class A
types
public R :: int;  
public U = R | B`R;
  
operations
public op : () ==> U -- Error reported: Operation returns unexpected type. Actual: B`R Expected: U
op () == return mk_B`R('a'); -- Value is not of the right type. Actual: B`R Expected: U  

end A

class B
types
public R :: char;
end B
~~~

Since `R`, defined in class `B`, is contained in `U` I would expect the spec to type check without errors. VDMJ, on the other hand, reports no type errors:

~~~
peter@peter-ubuntu:~/dev/vdm_tools$ java -jar vdmj-3.1.1.jar -vdmpp Bug.vdmpp -i
Parsed 2 classes in 0.148 secs. No syntax errors
Type checked 2 classes in 0.009 secs. No type errors
Initialized 2 classes in 0.031 secs. 
Interpreter started
> p new A().op()
= mk_B`R('a')
Executed in 0.033 secs. 
~~~

**Comment by nickbattle** [comment]
Confirmed that this is a problem. After a lot of visitor-hopping debugging (sigh), this comes down to a problem in PTypeComparator. The method compares the toStrings of the two types passed to it, but the toString of ARecordInvariantType doesn't include the "module name" (class name). So A`R and B`R are considered the same, and hence the union of them only contains one type.

We can try simply changing the toString for the type (or make sure that all types' toStrings are distinct), or we can try to make the PTypeComparator more intelligent. Changing the toString might have other side effects, depending on who else uses this method.

Changing the method as follows makes the example specification work:
```
	public int compare(PType o1, PType o2)
	{
		if (o1.equals(o2))
		{
			return 0;
		}
		else
		{
			return o1.hashCode() - o2.hashCode();   // Arbitrary ordering
		}
	}
```

**Comment by nickbattle** [comment]
Unfortunately the fix above causes all sorts of other problems. And I'm not sure why it works for this example here either - the equals() method of ARecordInvariantType just compares the toStrings of the types as well, so there's something strange going on.

**Comment by peterwvj** [comment]
Thanks for checking.

The `toString` of `ARecordInvariantType` is
~~~
public String toString()
{
  return "" + _name;
}
~~~

Since `name` is a `LexNameToken` effectively this is the same as invoking the `toString` of the `LexNameToken` which is

~~~
public String toString()
{
  return getFullName()
  + (typeQualifier == null ? "" : "("
  + Utils.listToString(typeQualifier) + ")");
}
~~~
But the `getFullName` yields "R" and so it does not really return the full name. This is a bit misleading. It is defined as:

~~~
public String getFullName()
{
  // Flat specifications have blank module names
  return (explicit ? module.length() > 0 ? module + "`" : "" : "") + name
     + (old ? "~" : ""); // NB. No qualifier
}
~~~

**Comment by peterwvj** [comment]
If I change `getFullName` to remove the explicit check I'm able to run the model
~~~
public String getFullName()
{
  // Flat specifications have blank module names
		return (module.length() > 0 ? module + "`" : "") + name
				+ (old ? "~" : ""); // NB. No qualifier
}
~~~

~~~
Parsed 2 classes in 0.111 secs. No syntax errors
Type checked 2 classes in 0.077 secs. No type errors
Initialized 2 classes in 0.108 secs. 
Interpreter started
> p new A().op()
= mk_B`R('a')
Executed in 0.052 secs. 
> 
~~~

But I'm not sure this is correct.

**Comment by ldcouto** [comment]
If this involves LexNameTokens, I'd like to point out https://github.com/overturetool/overture/issues/329

There are some long standing issues with names and name comparisons. Maybe this is a chance to improve on the issue?

**Comment by nickbattle** [comment]
The getFullName method seems to be used in lots of places. VDMJ doesn't have this method. It has a getExplicit(boolean) method which returns another LexNameToken with the explicit flag set or clear. This is quite complicated, unfortunately. The reason for the explicit flag is to do with when a name has explicitly been written with a module/class tag, like "mk_B`R(1,2,3)", or "obj.Super`op()". In these cases, the fact that the name was explicitly given can be important (ie. have a different meaning) to the case where the module/class is implicit (and set to whatever the containing module/class is).

I agree that getFullName *sounds* like it would give what we need, and if that could be changed that would make some sense. But a quick look at the places where getFullName is used suggests that several of them are depending on this not including the module/class tag. Of course those uses could be corrected, but there's quite a few of them (quite a few within the AST's toStrings).

Luis' point is good too. The LexNameToken has taken on too much functionality in VDMJ. It seemed sensible to qualify a name (of an overloaded function or operation) with a list of parameter types. But that introduces a dependency between the token name and the type system, which brings problems. Separating them again would be A Good Thing... but A Lot of Work too :(

**Comment by peterwvj** [comment]
Ah yes I see...

But if this is really what the `getFullName` is suppose to be doing then probably we should change the `toString` of the `ARecordInvariantType` to include the module name. It should only require a small tweak to the `overtureII.astv2.tostring` file... and it would not change the `LexNameToken` class.

It would not solve the problem we have with `LexNameToken` but it might fix this bug.

**Comment by nickbattle** [comment]
There are some other uses of the method, but yes, I agree with what you're suggesting. It doesn't make the LexNameToken "coupling" issue any worse, and we ought to be able to fix your bug. Want to give it a try? :)

**Comment by peterwvj** [comment]
Yes, sure I will try it and report :)

**Comment by peterwvj** [comment]
Okay so I did what we agreed would be a reasonable solution.

I should say that this fix affects many result files and it is impossible to inspect all of them. Those I checked seem correct.

Commit summary:

Actual fix:
https://github.com/overturetool/overture/commit/2584e7f6ce082c68f9201c84a1283ea59a6a3817

Bug regression test: 
https://github.com/overturetool/overture/commit/27de631ea8b92646dfc597dc320a3943d510fdb2

Update to result files:
https://github.com/overturetool/overture/commit/05e498bca825bb1277381be361384b8926f7653b

I pushed my work to the `issue-488` branch. Unless someone objects I will merged it into `pvj/main`
